streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -496,9 +496,15 @@ void commit(final boolean startNewTransaction, final Map<TopicPartition, Long> p
         }
 
         final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
+
         for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
             final TopicPartition partition = entry.getKey();
-            final long offset = entry.getValue() + 1;
+            Long offset = partitionGroup.headRecordOffset(partition);
+            if (offset == null) {
+                // this call should never block, because we know that we did process data for this partition
+                // and thus the consumer should have a valid local position that it can return immediately
+                offset = consumer.position(partition);
```

Comment:
We need to consider handling two exceptions that `consumer.position` may throw: `KafkaException` -> should be a fatal one; `TimeoutException` -> in this case we cannot commit, probably have to treat as fatal..

Reply:
Ack. We can wrap `KafkaException` as `StreamsException`. A `TimeoutException` should never happen (compare my comment -- let me know if you think the comment is incorrect) -- hence, we can rethrow `TimeoutException` as `IllegalStateException` to flag potential bugs.